sycl/doc/extensions/proposed/sycl_ext_oneapi_named_sub_group_sizes.asciidoc

```diff
@@ -0,0 +1,301 @@
+the notion of _named_ sub-group sizes, allowing developers to request a
+sub-group size that meets certain requirements at host compile-time and
+deferring the selection of a specific sub-group size until the kernel is
+compiled for a specific device.
+
+This extension also changes the default behavior of sub-groups in SYCL code to
+improve the out-of-the-box experience for new developers, without preventing
+experts and existing developers from requesting the existing compiler behavior.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_NAMED_SUB_GROUP_SIZES` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Changes to sub-group behavior
+
+Much of the behavior related to sub-groups in SYCL 2020 is
+implementation-defined. Different kernels may use different sub-group sizes,
+and even the same kernel may use different kernels on some devices (e.g. for
+different ND-range launch configurations).
+
+The extension introduces simpler behavior for sub-groups:
+
+- If no sub-group size property appears on a kernel or `SYCL_EXTERNAL`
+  function, the default behavior of an implementation must be to compile and
+  execute the kernel or function using a device's _primary_ sub-group size. The
+  primary sub-group size must be compatible with all core language features.
+
+- If a developer does not require a stable sub-group size across all kernels
+  and kernel launches, they can explicitly request an _automatic_ sub-group
+  size chosen by the implementation.
+
+- Implementations are free to provide mechanisms which override the default
+  sub-group behavior (e.g. via compiler flags), but developers must use this
+  mechanism explicitly in order to opt-in to any change in behavior.
+
+
+=== Device queries
+
+A new `info::device::primary_sub_group_size` device query is introduced to
+query a device's primary sub-group size.
+
+[%header,cols="1,5,5"]
+|===
+|Device Descriptor
+|Return Type
+|Description
+
+|`info::device::primary_sub_group_size`
+|`uint32_t`
+|Return a sub-group size supported by this device that is guaranteed to support
+ all core language features for the device.
+|===
+
+
+=== Properties
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+enum named_sub_group_size : uint32_t {
+  primary = /* implementation-defined */,
+  automatic = /* implementation-defined */,
+};
```

Comment:
Because this is an unscoped enum, the names `primary` and `automatic` are also in the `sycl::ext::oneapi::experimental` namespace.  I suspect that was not your intent?  Unfortunately, you cannot just make this a scoped enum because then the type of `primary` and `automatic` are no longer implicitly convertible to `uint32_t`.  Maybe this is better?

```
static constexpr uint32_t named_sub_group_primary = /*unspecified*/;
static constexpr uint32_t named_sub_group_automatic = /*unspecified*/;
```

Reply:
Ah, you're right...  I avoided making it a scoped enum precisely because of the `uint32_t` issue.

What about:
```c++
struct named_sub_group_size {
  static constexpr uint32_t primary = /* unspecified */;
  static constexpr uint32_t automatic = /* unspecified */;
};
```

I don't have a strong preference here, just floating this idea because I think it would allow the same syntax (`named_sub_group_size::primary`).

Reply:
Yes, that looks good too.  I don't have a strong preference between the two styles.

Reply:
I've gone with the `struct` approach in 23cca731eee054cb3c91c9fa4e469460b80dba4a, simply because it meant changing fewer lines.

I'd be happy to revisit this later.